### PR TITLE
Improve dataset generation and loading

### PR DESCRIPTION
**Pull request by mkrizmancic**
Changes:
- Add .gitignore
- Remove binary dataset files
- Fix dataset path
- Load a single batch if it fits in the memory - at least 2x faster than using the dataloader
- Cache the dataset generation using selected graph sizes and features. A dataset with a unique name is created for each combination of parameters and reused each time those parameters are specified. No more deleting and regenerating `data.pt` file for every change in the dataset.

### CODE REVIEW

File: algebraic_connectivity_notebook.ipynb [hunk 1]
```diff
@@ -112,7 +111,7 @@
     "    \"name\": \"MIDSdataset\",\n",
     "    \"selected_graphs\": str(selected_graph_sizes),\n",
     "    \"split\": 0.8,\n",
-    "    \"batch_size\": 6000,\n",
+    "    \"batch_size\": 1,\n",
```
Comment (mkrizmancic) [inline algebraic_connectivity_notebook.ipynb]: Batch size is now defined as a ratio (1 - fit the whole training dataset in one batch, 0.5 - put half of the data in one batch, etc.). The number of batches is ~ 1 / `batch_size`.

This should be easier than thinking about how many graphs are in the dataset and how many of them will fit in one batch.

We can revert this change if it's not desirable for some reason.

File: algebraic_connectivity_notebook.ipynb [hunk 2]
```diff
@@ -136,26 +135,27 @@
...
     "train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)\n",
-    "test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)\n",
+    "test_loader = DataLoader(test_dataset, batch_size=min(test_size, batch_size), shuffle=False)\n",
     "\n",
-    "train_batch = None\n",
-    "test_batch = None\n",
     "# If the whole dataset fits in memory, we can use the following lines to get a single large batch.\n",
-    "#train_batch = next(iter(train_loader))\n",
-    "#test_batch = next(iter(test_loader))\n",
+    "train_batch = next(iter(train_loader))\n",
```
Comment (mkrizmancic) [inline algebraic_connectivity_notebook.ipynb]: Taking an item from the loader is one of the slowest operations in the training process. If we have only one batch, we can do this once and speed up training at least 2x.

File: algebraic_connectivity_notebook.ipynb [hunk 3]
```diff
@@ -397,30 +402,53 @@
...
     "        #loss = check_accuracy_unsupervised(data, out, criterion).item()\n",
-    "        acc = check_accuracy(data,out)\n",
-    "    return loss, acc\n",
+    "        correct = 0\n",
+    "        start_slice = 0\n",
+    "        for i in range(len(data)):\n",
+    "            d = data[i]\n",
+    "            correct += check_accuracy(d, out[start_slice:start_slice+d.y.shape[0]])\n",
+    "            start_slice += d.y.shape[0]\n",
+    "    return loss, correct\n",
```
Comment (mkrizmancic) [inline algebraic_connectivity_notebook.ipynb]: Since we now have to handle both the whole dataset and a batch in the same way, this function is updated to take one graph at a time. `out` must be manually sliced using the number of nodes in `d` (we know that from the shape of target y).

File: algebraic_connectivity_notebook.ipynb [hunk 4]
```diff
@@ -441,10 +469,10 @@
     "    test_acc = 0\n",
     "    for epoch in range(1, num_epochs + 1):\n",
     "        # Perform one pass over the training set and then test on both sets.\n",
-    "        do_train(model, train_loader, optimizer, criterion)\n",
-    "        if epoch % 100 == 1 or epoch == 10000:\n",
-    "            train_loss, train_acc = do_test(model, train_dataset, criterion)\n",
-    "            test_loss, test_acc = do_test(model, test_dataset, criterion)\n",
+    "        train_loss = do_train(model, train_data_obj, optimizer, criterion)\n",
```
Comment (mkrizmancic) [inline algebraic_connectivity_notebook.ipynb]: For training, we calculate the loss in each epoch. Why not return it and monitor the training progress :)

File: algebraic_connectivity_notebook.ipynb [hunk 5]
```diff
@@ -441,10 +469,10 @@
...
+    "        if epoch % 100 == 0 or epoch == num_epochs:\n",
```
Comment (mkrizmancic) [inline algebraic_connectivity_notebook.ipynb]: Skips evaluation in the first epoch. Add an additional condition if this is desired.

File: pyg_dataset.py [hunk 6]
```diff
@@ -54,12 +55,11 @@ def processed_file_names(self):
...
-        #       We could come up with a namig scheme that will differentiate
-        #       which graph families (and/or sizes) and features were used to
-        #       generate the dataset. This way, we could detect changes and
-        #       reprocess the dataset when needed.
-        return ["data.pt"]
+        dataset_props = json.dumps([self.loader.selection, self.features])
+        sha256_hash = hashlib.sha256(dataset_props.encode("utf-8")).digest()
+        hash_string = base64.urlsafe_b64encode(sha256_hash).decode("utf-8")[:10]
+
+        return [f"data_{hash_string}.pt"]
```
Comment (mkrizmancic) [inline pyg_dataset.py:62]: Makes a hash of selected graph sizes and features, and generates a unique file name. If the same parameters are given in another run, and if the file exists, it will be loaded. Otherwise, the dataset will be automatically regenerated.

File: pyg_dataset.py [hunk 7]
```diff
@@ -85,39 +85,26 @@ def process(self):
...
+    # Define features in use.
+    feature_functions = {
+        "degree": lambda g: {n: float(g.degree(n)) for n in g.nodes()},
+        "degree_centrality": nx.degree_centrality,
+        "random": lambda g: nx.random_layout(g, seed=np.random),
+        "avg_neighbor_degree": nx.average_neighbor_degree,
+        "closeness_centrality": nx.closeness_centrality,
+        "number_of_nodes": lambda g: [nx.number_of_nodes(g)] * nx.number_of_nodes(g)
+    }
+
```
Comment (mkrizmancic) [inline pyg_dataset.py:98]: I had to remove this from the function to have access to the list of features.
